Implementation of reverse type reference lookup

trylek opened the pull request:
Based on JanK's suggestion, I have implemented the basic logic
for reverse type reference lookup. I have demonstrated basic
functionality of this feature by adding two delegate tests involving
Action and Func<char, char>.

I also had to slightly refactor type signature emitter - my
initial implementation was heavily dependent on parsing an existing
metadata signature but in the resolution process we can hit synthetic
types that don't have such a metadata representation, most notably
for generics.

For this reason I believe we need custom recursive descent into the
type tree and this change includes the initial implementations. For
now I'm keeping the generic signature decoder as it's still used by
method signatures but my current thinking is that it will also become
unneeded at some point and we'll be better of with a custom
implementation matching our exact needs.

Thanks

Tomas

nattress commented on src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs:
```diff
@@ -767,25 +768,54 @@ public ISymbolNode DelegateCtor(TypeDesc delegateType, MethodDesc targetMethod,
+                    foreach (TypeReferenceHandle typeRefHandle in metadataReader.TypeReferences)
+                    {
+                        TypeDesc resolvedType = (TypeDesc)module.GetObject((EntityHandle)typeRefHandle);
+                        _typeToRefTokens[resolvedType] = new ModuleToken(module, (mdToken)MetadataTokens.GetToken(typeRefHandle));
```
Thinking ahead to tests that will compare r2rdump output from crossgen and ILCompiler, can we ensure we'll pick the same answer in situation where multiple assemblies in the input file set have typerefs to the same type? Here, it's defined by the order of assemblies passed in to ILCompiler. Ideally it would be stabilized by, for example, assembly import ordering in the main input assembly.

trylek replied:
Agreed. Poking around CoreCLR, it seems to me that right now Crossgen doesn't actually have any provisions like this as it likely only supports single-file version bubble in R2R compilation mode. I think we'll need to add at least some rudimentary provisions for multi-module bubble to the legacy CoreCLR Crossgen as according to JanK it should be fairly straightforward and it's the easiest way to obtain code comparisons we can use to monitor CQ.

nattress commented on src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs:
```diff
@@ -767,25 +768,54 @@ public ISymbolNode DelegateCtor(TypeDesc delegateType, MethodDesc targetMethod,
+        /// </summary>
+        private Dictionary<TypeDesc, ModuleToken> _typeToRefTokens = null;
+
+        public ModuleToken GetModuleTokenForType(EcmaType type)
```
This method is currently only called from SignatureBuilder, where we have `_contextModule`. Should the context module be passed along here so that we get a typeref from the correct "point of view". I think we'll end up generating invalid signatures as-is. For example, if assemblies A and B have a typeref to System.Object and we're building a signature for A (A is the `_contextModule`), we don't want to put a typeref from B in the signature since it'll point at the wrong type most likely.

trylek replied:
I think we should be able to use a typeref relative to a different module than the one we're compiling right now, we just need to use the 'module override' flag in the signature. I admit I haven't tried this out yet though (in part because the relevant infrastructure isn't yet in place). Based on reading the CoreCLR runtime code I assume we'll somehow need to emit additional fixup cells representing the additional modules within the versioning bubble and encode their indices in the signature.

MichalStrehovsky commented on src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs:
```diff
@@ -767,25 +768,54 @@ public ISymbolNode DelegateCtor(TypeDesc delegateType, MethodDesc targetMethod,
+        /// gets lazily initialized with all input module references the first time the reverse lookup is
+        /// needed.
+        /// </summary>
+        private Dictionary<TypeDesc, ModuleToken> _typeToRefTokens = null;
```
Wouldn't it be easier to keep track of this in the JitInterface? RyuJIT is going to call `resolveToken` to resolve all tokens. We could just stash away those that have `IsDelegate == true` and when JitInterface calls `NodeFactory.DelegateCtor`, we could just find the token in the map and provide it as a parameter there. The fix would be more localized to the place that has the problem.

That way we avoid having to go over all the TypeRef tokens that is just asking for extra trouble (users love to ship their code with incomplete/unresolvable/junk type references).

trylek replied:
Hmm, I just tried this out but unfortunately it doesn't work. We don't only need the tokens for delegate types, we need them in all sorts of places around the signature encoding when we don't have a token readily available which is the case most of the time for composite signatures like instantiations.

In particular, just using pre-recorded calls to resolveToken I'm unable to resolve List`1 because that never enters resolveToken by itself even though many methods on an instantiation of the type do.

Gracefully ignoring unresolvable references is probably a worthy goal; please let me also remind you that in the expected typical user case the "external references" involve "our" framework code. Other than that, the current implementation is more or less an improved version of

https://github.com/dotnet/coreclr/blob/61146b5c5851698e113e936d4e4b51b628095f27/src/vm/ceeload.cpp#L5872

trylek commented:
I have removed the reverse lookup per Michal's suggestion and I added initial support for unboxing stubs and shared generics. I have also fixed one bug in the X64 assembly emitter. @MichalStrehovsky and @nattress, PTAL. I'll need to sync up with Simon regarding merging in, I'm prepared to wait for Simon and bite the bullet of the NodeFactory conflict.

Thanks

Tomas

MichalStrehovsky commented on src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRun/SignatureBuilder.cs:
```diff
@@ -179,149 +179,181 @@ internal void EmitElementType(CorElementType elementType)
 
-            // TODO: module override for tokens with external context
-            EmitElementType(elementType);
+                case TypeFlags.ValueType:
```
Nit: seems like `ValueType` and `Nullable` cases could be merged into one

MichalStrehovsky commented on src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRun/Target_X64/DelayLoadHelperThunk.cs:
```diff
@@ -16,8 +16,20 @@ public partial class DelayLoadHelperThunk
-            instructionEncoder.EmitLEAQ(X64.Register.RAX, _instanceCell);
+            if (_isVirtualStubDispatchCell)
+            {
+                // lea rax, r11 - this is the most general case as the value of R11 also propagates
```
The `lea` part in the comment doesn't match what the code does

MichalStrehovsky commented on src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRun/Target_X64/DelayLoadHelperThunk.cs:
```diff
@@ -16,8 +16,20 @@ public partial class DelayLoadHelperThunk
+                // lea rax, r11 - this is the most general case as the value of R11 also propagates
+                // to the new method after the indirection cell has been updated so the cell content
+                // can be repeatedly modified as needed during virtual / interface dispatch.
+                instructionEncoder.EmitMOV(X64.Register.RAX, X64.Register.R11);
```
Nit: I would add `using ILCompiler.DependencyAnalysis.X64` at the top. This file is X64 only, so we don't need to qualify what kind of `Register` this is.

MichalStrehovsky commented on src/JitInterface/src/CorInfoImpl.cs:
```diff
@@ -1114,12 +1114,16 @@ private void resolveToken(ref CORINFO_RESOLVED_TOKEN pResolvedToken)
             pResolvedToken.hMethod = null;
             pResolvedToken.hField = null;
-
+            
```
Nit: unnecessary whitespace change

MichalStrehovsky commented on src/JitInterface/src/CorInfoImpl.cs:
```diff
@@ -1114,12 +1114,16 @@ private void resolveToken(ref CORINFO_RESOLVED_TOKEN pResolvedToken)
                 pResolvedToken.hClass = ObjectToHandle(method.OwningType);
+
+#if READYTORUN
+            _compilation.NodeFactory.AddModuleTokenForMethod(method, new ModuleToken(_tokenContext, (mdToken)pResolvedToken.token));
```
I was hoping we would be stashing this information in a map that is local to `CorInfoImpl` and use the mapping when `CorInfoImpl` needs to talk to an API on `NodeFactory` that needs the token, but we don't have it at that point (like in the delegate construction case that prompted this). Are there other places where this would be insufficient? I'm worried that storing this in a global place like the `NodeFactory` doesn't actually solve the problem of us being able to reference the same method through different tokens, and which token to use when.

This is an improvement in the sense that we don't have to reverse scan the ECMA metadata table, but it doesn't go as far as I hoped it would.

MichalStrehovsky commented on src/JitInterface/src/CorInfoImpl.cs:
```diff
@@ -3076,13 +3092,13 @@ private void getCallInfo(ref CORINFO_RESOLVED_TOKEN pResolvedToken, CORINFO_RESO
                         _compilation.NodeFactory.InterfaceDispatchCell(targetMethod
-#if !SUPPORT_JIT
+#endif // READYTORUN
+#if !SUPPORT_JIT || READYTORUN
```
Why do we need the `|| READYTORUN` part?

MichalStrehovsky replied:
> Why do we need the || READYTORUN part in CorInfoImpl around line
3100? To let us pass around call site names to interface dispatch
cells - as interface dispatch cells are per call site even though
they might reference the same interface method, we need some way
to distinguish their symbols as we validate symbol uniqueness to
maintain output consistency.

Does R2R compiler define `SUPPORT_JIT`? `SUPPORT_JIT` should not be defined - it indicates CorInfoImpl behaviors needed for the JIT prototype (System.Private.Jit). Once R2R stops defining it, this `if` check will be true (because the condition is "not experimental JIT prototype").

nattress commented on src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRun/ModuleToken.cs:
```diff
@@ -24,6 +25,19 @@ public ModuleToken(EcmaModule module, mdToken token)
+            Token = (mdToken)MetadataTokens.GetToken(handle);
+        }
+
+        public bool IsNull => Module == null;
```
I'm a little unsure how to use this class correctly. Is it invalid to call `Handle` if `IsNull == true`?

trylek commented:
@nattress Sadly I can no longer reply to the particular comment after I pushed the rebased change against your recent merge (please let me know whether there's any more graceful way of doing rebases) but in general, when ModuleToken.IsNull, its Handle is invalid. If you're saying that perhaps we should add an assertion check that we don't inadvertently use the Handle of a null ModuleToken, that sounds like a reasonable suggestion to me.

nattress commented:
Yes, an assertion to guide the way would be great thank you

trylek commented:
@MichalStrehovsky Ad VALUETYPE vs. NULLABLE - interesting, when I reviewed the code, at first I thought this was just a copy & paste typo but apparently it isn't. Thanks for the suggestion, I'm unifying these in my new commit!

trylek commented:
@MichalStrehovsky I have addressed part of your PR feedback in my latest commit and I tried to explain my rationale for the rest in my commit message. I'm looking forward to your additional feedback.

Thanks

Tomas
